<a href="https://colab.research.google.com/github/musawenkosikhulu/-Linear-Regression/blob/main/Linear_Regression_Prac_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Linear Regression Prac 2

This notebook will make use of Scikit-learn's Linear Regression class. The documentation is found here: https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html

Credits:

The notebooks on linear regression are modifications of various sources which include:
https://scikit-learn.org/stable/auto_examples/linear_model/plot_ols.html
https://medium.com/analytics-vidhya/linear-regression-using-python-ce21aa90ade6
https://www.kdnuggets.com/2019/03/beginners-guide-linear-regression-python-scikit-learn.html/2

## Various Python imports

In [ ]:
import matplotlib.pyplot as plt
import numpy as np
from sklearn import datasets
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error
from sklearn.model_selection import train_test_split

## Load the dataset

In [ ]:
diabetes = datasets.load_diabetes()

## Take a look at what has been downloaded

In [ ]:
diabetes

{'DESCR': '.. _diabetes_dataset:\n\nDiabetes dataset\n----------------\n\nTen baseline variables, age, sex, body mass index, average blood\npressure, and six blood serum measurements were obtained for each of n =\n442 diabetes patients, as well as the response of interest, a\nquantitative measure of disease progression one year after baseline.\n\n**Data Set Characteristics:**\n\n  :Number of Instances: 442\n\n  :Number of Attributes: First 10 columns are numeric predictive values\n\n  :Target: Column 11 is a quantitative measure of disease progression one year after baseline\n\n  :Attribute Information:\n      - age     age in years\n      - sex\n      - bmi     body mass index\n      - bp      average blood pressure\n      - s1      tc, total serum cholesterol\n      - s2      ldl, low-density lipoproteins\n      - s3      hdl, high-density lipoproteins\n      - s4      tch, total cholesterol / HDL\n      - s5      ltg, possibly log of serum triglycerides level\n      - s6      glu, b

## Print out the feature names

In [ ]:
diabetes['feature_names']

['age', 'sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6']

## Take a look at the shape of the data

In [ ]:
diabetes['data'].shape

(442, 10)

## Take a look at the shape of the targets (what we are predicting for)

In [ ]:
diabetes['target'].shape

(442,)

## Use all training features

In this notebook we are interested in creating a multiple variable regression model. So in this case, we will use all variables.

Typically, features are denoted as X and targets as Y (some authors will use 'y' instead).

In [ ]:
X = diabetes.data
y = diabetes['target']

## Check the shape!

In [ ]:
X.shape

(442, 10)

In [ ]:
y.shape

(442,)

## Split the data into training and testing sets

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=4278) # 70% training and 30% test

## Check the training features shape

In [ ]:
X_train.shape

(309, 10)

## Check the testing features shape

In [ ]:
X_test.shape

(133, 10)

## Define a linear regression model

In [ ]:
model = LinearRegression()

## Look at what has just being initialised

In [ ]:
model

LinearRegression()

## Train the model/fit the model

In [ ]:
model.fit(X_train, y_train)

LinearRegression()

## Print the coefficients

In [ ]:
model.coef_

array([ -48.21538346, -208.21759667,  538.21407795,  317.59757521,
       -412.84092159,  192.07650332,   -7.83180383,  210.28615811,
        608.10859302,   49.99113799])

## Print the intercept

In [ ]:
model.intercept_

152.05315468620535

## Task: can you write out the model given the above outputs?
y = 152 -48d1 -208d2 + 538d3 + 317d4 -412d5 +192d6 -7d7 + 210d8 + 608d9 + 49d10
...

## Apply the model to the testing features and get the predictions

In [ ]:
Y_pred = model.predict(X_test)

## Determine the performance of the model on the testing data

There are various options, but commonly used one include:

* mean squared error
* mean absolute error
* root mean squared error

There can be implemented using scikit-learn.

### Mean squared error

In [ ]:
mean_squared_error(y_test, Y_pred)

2938.633931374276

### Mean absolute error

In [ ]:
mean_absolute_error(y_test, Y_pred)

45.14810055653443

### Root mean squared error

In [ ]:
np.sqrt(mean_squared_error(y_test, Y_pred))

54.209168333172904

## Task: is the performance better than using a single variable?

Yes